src/python/pants/util/strutil.py

```diff
@@ -353,10 +353,8 @@ def default(self, o):
             return dict(o)
         if isinstance(o, abc.Sequence):
             return list(o)
-        try:
-            return super().default(o)
-        except TypeError:
-            return str(o)
+        return super().default(o)
```

Comment:
I disagree with this change.

Reply:
I could imagine not cherry-picking the fallback-removal portion though, if the concern is breaking cases which happened to be safe...?

Reply:
```diff
diff --git a/src/python/pants/backend/docker/utils.py b/src/python/pants/backend/docker/utils.py
index ec09791636..27b1fc1b0a 100644
--- a/src/python/pants/backend/docker/utils.py
+++ b/src/python/pants/backend/docker/utils.py
@@ -186,7 +186,7 @@ class JsonEncoder(json.JSONEncoder):
         try:
             return super().default(o)
         except TypeError:
-            return str(o)
+            raise
 
 
 json_dumps = partial(
```
```shell
pants test src/python/pants/backend/docker/util_rules/docker_build_context_test.py -- -k hash
```

```
     E       pants.engine.internals.scheduler.ExecutionError: 1 Exception encountered:
E
E       Engine traceback:
E         in select
E           ..
E         in pants.backend.docker.util_rules.docker_build_context.create_docker_build_context
E           ..
E
E       Traceback (most recent call last):
E         File "/private/var/folders/y2/6t5bbk554wv5h6w_qlgpcpnm0000gr/T/pants-sandbox-tqp3kC/src/python/pants/engine/internals/selectors.py", line 626, in native_engine_generator_send
E           res = rule.send(arg) if err is None else rule.throw(throw or err)
E         File "/private/var/folders/y2/6t5bbk554wv5h6w_qlgpcpnm0000gr/T/pants-sandbox-tqp3kC/src/python/pants/backend/docker/util_rules/docker_build_context.py", line 376, in create_docker_build_context
E           return DockerBuildContext.create(
E         File "/private/var/folders/y2/6t5bbk554wv5h6w_qlgpcpnm0000gr/T/pants-sandbox-tqp3kC/src/python/pants/backend/docker/util_rules/docker_build_context.py", line 133, in create
E           "hash": get_hash((build_args, build_env, snapshot.digest)).hexdigest(),
E         File "/private/var/folders/y2/6t5bbk554wv5h6w_qlgpcpnm0000gr/T/pants-sandbox-tqp3kC/src/python/pants/backend/docker/utils.py", line 198, in get_hash
E           return hashlib.new(name, json_dumps(value).encode("utf-8"))
E         File "/Users/andreas.stenius/.pyenv/versions/3.9.13/lib/python3.9/json/__init__.py", line 234, in dumps
E           return cls(
E         File "/Users/andreas.stenius/.pyenv/versions/3.9.13/lib/python3.9/json/encoder.py", line 199, in encode
E           chunks = self.iterencode(o, _one_shot=True)
E         File "/Users/andreas.stenius/.pyenv/versions/3.9.13/lib/python3.9/json/encoder.py", line 257, in iterencode
E           return _iterencode(o, 0)
E         File "/private/var/folders/y2/6t5bbk554wv5h6w_qlgpcpnm0000gr/T/pants-sandbox-tqp3kC/src/python/pants/backend/docker/utils.py", line 187, in default
E           return super().default(o)
E         File "/Users/andreas.stenius/.pyenv/versions/3.9.13/lib/python3.9/json/encoder.py", line 179, in default
E           raise TypeError(f'Object of type {o.__class__.__name__} '
E       TypeError: Object of type DockerBuildArgs is not JSON serializable
```

Reply:
> Here's one huge gigantic ptifall for calling this a "stable" hash:

I think we may want to establish our definition of what a stable hash means here first. As I wouldn't think the hash for `object()` to be stable in the first place. Granted there are certainly valid instances out there that presents similar volatile data that we might want to consider as stable for input purposes for the hash but right now it's all very hypothetical. Why not have the premise that `stable_hash()` is only guaranteed to produce valid (i.e. stable) hash values for stable inputs. Instances that does not present a stable string representation could then be deemed unstable for such purposes.

Reply:
> I think that likely `Sequence` is too narrowing there. `AbstractOrderedSet` is not a `Sequence` but is `Collection`. So perhaps thats the definer.

Cool. OK I'm convinced, as long as the docker tests pass we should be fine. And there is a benefit to be alerted about potentially unstable hashes for unhandled types/protocols.

Reply:
I ended up special casing `(Frozen)OrderedSet`. normal `set`s are `Collection`s but aren't stable so that was unsafe.